openassessment/fileupload/api.py

```diff
@@ -29,3 +37,299 @@ def remove_file(key):
+        """
+        Returns:
+            A list of FileUpload objects associated with an instance of an O.A. Block.
+        """
+        descriptions, names, sizes = self._get_metadata_from_block()
+        return self._file_uploads_from_list_fields(descriptions, names, sizes)
+
+    def append_uploads(self, *new_uploads):
+        """
+        Given lists of new file metadata, write the new metadata to our stored file metadata fields
+
+        Args:
+            descriptions_to_add: a list of file descriptions
+            names_to_add: a list of file names
+            sizes_to_add: a list of file sizes as integers
+
+        Returns: newly updated file metadata fields
+        """
+        required_keys = ('description', 'name', 'size')
+        try:
+            (
+                descriptions_to_add,
+                names_to_add,
+                sizes_to_add,
+            ) = self._dicts_to_key_lists(new_uploads, required_keys)
+        except FileUploadError:
+            raise
+
+        existing_file_descriptions, existing_file_names, existing_file_sizes = self._get_metadata_from_block()
+
+        new_descriptions = existing_file_descriptions + descriptions_to_add
+        self._set_file_descriptions(new_descriptions)
+
+        new_names = existing_file_names + names_to_add
+        self._set_file_names(new_names)
+
+        new_sizes = existing_file_sizes + sizes_to_add
+        self._set_file_sizes(new_sizes)
+
+        return self._file_uploads_from_list_fields(new_descriptions, new_names, new_sizes)
+
+    def delete_upload(self, index):
+        """
+        Given a file index to remove, null out its metadata in our stored file metadata fields
+
+        Args:
+            index: file index to remove
+
+        Returns: newly updated FileUpload records.
+        """
+        file_key = get_student_file_key(self.block.get_student_item_dict(), index)
+        remove_file(file_key)
+
+        stored_file_descriptions, stored_file_names, stored_file_sizes = self._get_metadata_from_block()
+
+        stored_file_descriptions[index] = None
+        self._set_file_descriptions(stored_file_descriptions)
+
+        stored_file_names[index] = None
+        self._set_file_names(stored_file_names)
+
+        stored_file_sizes[index] = 0
+        self._set_file_sizes(stored_file_sizes)
+
+    def _get_metadata_from_block(self):
+        descriptions = self._get_file_descriptions()
+        names = self._get_file_names(descriptions)
+        sizes = self._get_file_sizes(descriptions)
+        return descriptions, names, sizes
+
+    def _file_uploads_from_list_fields(self, descriptions, names, sizes, include_deleted=False):
+        """
+        Given file upload data as list fields, return a list of FileUploads constructed from those fields
+        """
+        file_fields_by_key = {
+            'name': names,
+            'description': descriptions,
+            'size': sizes,
+        }
+
+        if not descriptions:
```

Comment:
Just checking that this is required to handle the old boundary case - i.e: we can't throw an exception if descriptions are null?

Reply:
Hmm, I don't think so, but I'll make sure that we have a unit test that covers that condition.

Reply:
👍 Added a unit test and fixed a bug.